// node.class/test/build/compiled_expected/node_modules/Class.class.js
var assert = require('assert');

var Extends = require('SuperClass.class');
var Class = {
    "init": function(arg1, arg2) {
        this.isClassInit = !this.isClassInit;
        this.__args = [arg1, arg2];
        assert.ok(typeof this.Super === "function");
        this.Super("different argument 1", "different argument 2");
        assert.ok(this.Super instanceof Extends);
    },
    "__args": null,
    "getArgs": function() {
        return this.__args.concat(this.Super.getArgs());
    },
    "assertProperties": function() {
        assert.ok(this.init === undefined);
        assert.ok(this.__args instanceof Array);
        assert.ok(typeof this.getArgs === "function");
        assert.ok(typeof this.assertProperties === "function");
        assert.ok(typeof this.isClassInit === "boolean");
        assert.ok(this.setIsClassInit === undefined);
        assert.ok(this.getIsClassInit === undefined);
        assert.ok(this.setIsSuperClassInit === undefined);
        assert.ok(this.getIsSuperClassInit === undefined);
        assert.ok(Constructor.staticString === "static");
        assert.ok(typeof this.abstract === "function");
        assert.ok(typeof this.showThis === "function");
        assert.ok(typeof this.getPrivateStuffFromSuper === "function");
        assert.ok(this.Super.__args === undefined);
        assert.ok(typeof this.Super.getArgs === "function");
        assert.ok(typeof this.Super.assertProperties === "function");
        assert.ok(typeof this.Super.getIsSuperClassInit === "function");
        assert.ok(typeof this.Super.setIsSuperClassInit === "function");
        assert.ok(typeof this.Super.getIsSuperSuperClassInit === "function");
        assert.ok(typeof this.Super.setIsSuperSuperClassInit === "function");
        assert.ok(typeof this.Super.setProtectedStuff === "function");
        assert.ok(typeof this.Super.getProtectedStuff === "function");
        assert.ok(typeof this.Super.callAbstract === "function");
        assert.ok(typeof this.Super.getPrivateStuff === "function");
        assert.ok(typeof this.Super.setPrivateStuff === "function");
        assert.ok(typeof this.Super.showThis === "function");
        assert.ok(this.Instance instanceof Constructor);
        this.Super.assertProperties();
    },
    "isClassInit": false,
    "$staticString": "static",
    "abstract": function() {
        return "i am not abstract anymore";
    },
    "showThis": function() {
        console.log(this);
        this.Super.showThis();
    },
    "getPrivateStuffFromSuper": function() {
        return this.Super.getPrivateStuff();
    }
};

// example
var Module = this;
var Constructor = function() {
    var tmp,
        implChildAbstracts,
        implAbstracts,
        exposeProtected,
        exposeNothing,
        Properties,
        Instance,
        key;

    if(Constructor.$) {
        tmp = Constructor.$;
        Instance = tmp.Instance;
        implChildAbstracts = tmp.implChildAbstracts;
        exposeProtected = tmp.exposeProtected;
        exposeNothing = tmp.exposeNothing;
    }
    if(!exposeNothing) {
        if(!implChildAbstracts) {
            implChildAbstracts = {};
        }

        Properties = {};

        if (Instance === undefined) {
            Instance = this;
        }
        Properties.isClassInit = false;
        Properties.__args = null;
        Properties.getArgs = Class.getArgs.bind(Properties);
        Properties.abstract = Class.abstract.bind(Properties);
        Properties.showThis = Class.showThis.bind(Properties);
        Properties.assertProperties = Class.assertProperties.bind(Properties);
        Properties.getPrivateStuffFromSuper = Class.getPrivateStuffFromSuper.bind(Properties);
        Properties.Instance = Instance;

        this.abstract = Class.abstract.bind(Properties);
        this.showThis = Class.showThis.bind(Properties);
        this.getArgs = Class.getArgs.bind(Properties);
        this.assertProperties = Class.assertProperties.bind(Properties);
        this.setIsClassInit = function(value) {
            Properties.isClassInit = value;
        };
        this.getIsClassInit = function() {
            return Properties.isClassInit;
        };
        this.getPrivateStuffFromSuper = Class.getPrivateStuffFromSuper.bind(Properties);

        implAbstracts = {};
        implAbstracts.abstract = Class.abstract.bind(Properties);

        if(implAbstracts) {
            for(key in implChildAbstracts) {
                implAbstracts[key] = implChildAbstracts[key];
                Properties[key] = implChildAbstracts[key];
            }
        } else {
            implAbstracts = implChildAbstracts;
            for(key in implChildAbstracts) {
                Properties[key] = implChildAbstracts[key];
            }
        }

        Properties.Super = function() {
            Extends.$ = {
                "exposeProtected": true,
                "implChildAbstracts": implAbstracts,
                "Instance": Instance
            };
            Properties.Super = Extends.$construct(arguments);
        };

        Class.init.apply(Properties, arguments);

        if(typeof Properties.Super === 'function') {
            Properties.Super.apply(Module, arguments);
        }

        if(exposeProtected) {
        }

        this.callAbstract = Properties.Super.callAbstract;
        this.setIsSuperClassInit = Properties.Super.setIsSuperClassInit;
        this.getIsSuperClassInit = Properties.Super.getIsSuperClassInit;
        this.getIsSuperSuperClassInit = Properties.Super.getIsSuperSuperClassInit;
        this.setIsSuperSuperClassInit = Properties.Super.setIsSuperSuperClassInit;
        this.setPrivateStuff = Properties.Super.setPrivateStuff;
    }

    delete Constructor.$;
};
Constructor.staticString = "static";

Constructor.$init = function() {
    if(Extends.$init) {
        Extends.$init();
    }
    Extends.$ = {
        "exposeNothing": true
    };
    Constructor.prototype = new Extends();

    delete Constructor.$init;
};

Constructor.$construct = function(args) {
    var Instance = function Instance() {
        return Constructor.apply(this, args);
    }
    Instance.prototype = Constructor.prototype;
    return new Instance();
};

module.exports = Constructor;
