// node.class/test/build/compiled_expected/node_modules/SuperSuperClass.class.js
var assert = require('assert'),
    Class;

Class = {
    "init": function(arg1, arg2) {
        Class = require('./Class.class');
        this.isSuperSuperClassInit = !this.isSuperSuperClassInit;
        this.__args = [arg1, arg2];
    },
    "__args": null,
    "getArgs": function() {
        return this.__args;
    },
    "isSuperSuperClassInit": false,
    "assertProperties": function() {
        assert.ok(this.init === undefined);
        assert.ok(this.__args instanceof Array);
        assert.ok(typeof this.__privateStuff === "number");
        assert.ok(typeof this.getArgs === "function");
        assert.ok(typeof this.assertProperties === "function");
        assert.ok(typeof this.isSuperSuperClassInit === "boolean");
        assert.ok(typeof this.abstract === "function");
        assert.ok(this["?abstract"] === undefined);
        assert.ok(typeof this.setPrivateStuff === "function");
        assert.ok(typeof this._getPrivateStuff === "function");
        assert.ok(typeof this.showThis === "function");
        assert.ok(typeof this.callAbstract === "function");
        assert.ok(this.Super === undefined);
        assert.ok(this.Instance instanceof require('./Class.class.js'));
    },
    "?abstract": Function,
    "callAbstract": function() {
        return this.abstract();
    },
    "__privateStuff": 23,
    "_getPrivateStuff": function() {
        return this.__privateStuff;
    },
    "setPrivateStuff": function(value) {
        this.__privateStuff = value;
    },
    "showThis": function() {
        console.log(this);
    },
    "getInstance": function() {
        return this.Instance;
    }
};

// example
var Module = this;
var Constructor = function() {
    var tmp,
        implChildAbstracts,
        implAbstracts,
        exposeProtected,
        exposeNothing,
        Properties,
        Instance,
        key;

    if(Constructor.$) {
        tmp = Constructor.$;
        Instance = tmp.Instance;
        implChildAbstracts = tmp.implChildAbstracts;
        exposeProtected = tmp.exposeProtected;
        exposeNothing = tmp.exposeNothing;
    }

    if(!exposeNothing) {
        Properties = {};

        if (Instance === undefined) {
            Instance = this;
        }
        Properties.isSuperSuperClassInit = false;
        Properties.__args = null;
        Properties.getArgs = Class.getArgs.bind(Properties);
        Properties._getPrivateStuff = Class._getPrivateStuff.bind(Properties);
        Properties.setPrivateStuff = Class.setPrivateStuff.bind(Properties);
        Properties.__privateStuff = 23;
        Properties.assertProperties = Class.assertProperties.bind(Properties);
        Properties.showThis = Class.showThis.bind(Properties);
        Properties.callAbstract = Class.callAbstract.bind(Properties);
        Properties.getInstance = Class.getInstance.bind(Properties);
        Properties.Instance = Instance;

        this.showThis = Class.showThis.bind(Properties);
        this.getArgs = Class.getArgs.bind(Properties);
        this.setIsSuperSuperClassInit = function(value) {
            Properties.isSuperSuperClassInit = value;
        };
        this.getIsSuperSuperClassInit = function() {
            return Properties.isSuperSuperClassInit;
        };
        this.assertProperties = Class.assertProperties.bind(Properties);
        this.callAbstract = Class.callAbstract.bind(Properties);
        this.setPrivateStuff = Class.setPrivateStuff.bind(Properties);

        if(implAbstracts) {
            for(key in implChildAbstracts) {
                implAbstracts[key] = implChildAbstracts[key];
                Properties[key] = implChildAbstracts[key];
            }
        } else {
            implAbstracts = implChildAbstracts;
            for(key in implChildAbstracts) {
                Properties[key] = implChildAbstracts[key];
            }
        }

        if(exposeProtected) {
            this.getPrivateStuff = Class._getPrivateStuff.bind(Properties);
        }

        Class.init.apply(Properties, arguments);
    }

    delete Constructor.$;
};
Constructor.$construct = function(args) {
    var Instance = function Instance() {
        return Constructor.apply(this, args);
    }
    Instance.prototype = Constructor.prototype;
    return new Instance();
};

module.exports = Constructor;
